### DISCUSSION TITLE
Improved Table input text editor + variable support

### DISCUSSION THREAD

**Discussion by rvirgolireply** [apache/hop#4670] [Ideas]
Hello,
I've been using Hop for quite a while and I feel like the text editor in the table input (and in the formula as well) might be improved by:
- being able to set tabulation to 4 spaces by default
- making it a proper text editor, with keyword highlights and stuff
- making it capable to use variables.
I find the last point in particular to be very critical, because it is impossible to get output fields while developing the pipeline from the table input if there's a variable in it even with "replace variables" checked.
Not sure if there's already a proper issue on these topics, at least I couldn't find it.

What do you think? 
Thanks!

**Comment by bamaer** [comment]
Hi @rvirgolireply, 

- 4 spaces tabulation: good point, feel free to create a feature request ticket 
- Syntax highlighting will be available in the upcoming 2.11 release. 
- Variables: this works, but the issue you're having could be because your variables don't have a (default) value. Can you reproduce this if you specify your variable as a pipeline parameter with a default value? 

Hope that helps, 
Bart

**Comment by rvirgolireply** [comment]
Thanks!
Didn't know that I could set a parameter, this is great news! Also, did I miss this info in the docs?
Regarding the first point, I'll open the feature request :)

**Reply by bamaer** [reply]
parameters are [documented](https://hop.apache.org//manual/latest/variables.html#_parameters), but just like a lot of our documentation, it may not always be where you expect it. 
Where did you expect it? We could add a pointer to parameter documentation in the pipeline and workflow docs pages, but I guess the search on the hop website is your best friend.

**Reply by hansva** [reply]
it's in the docs but maybe not prominent enough.
we mention it [here](https://hop.apache.org//manual/latest/variables.html#_how_can_i_define_variables) and [here](https://hop.apache.org//manual/latest/pipeline/create-pipeline.html#_pipeline_properties)

**Reply by rvirgolireply** [reply]
Oh okay, I read those sections but it isn't clear that if you don't add those then you cannot get previews of subsequent transforms.
In other words, while developing a pipeline it is almost mandatory to set the parameters, otherwise it's impossible to see output fields from the transform that uses such variable.
Thanks guys!

**Reply by hansva** [reply]
Thanks for that feedback, feel free to create a documentation ticket for it too so we don't lose the task.
